Deprecate engines.

paulidale opened the pull request:
Deprecation of engine support in 3.0.

This deals with the public APIs and their internal use.  It doesn't not address the removal of the engine support code itself.

- [x] documentation is added or updated
- [ ] tests are added or updated

paulidale commented:
This remains very rough and definitely work in progress.  The _no-deprecated_ build is failing due to symbol problems in `util/libcrypto.num`.  `make update` is adding `DEPRECATE_3_0` tags in addition to the `DEPRECATEDIN_3_0` ones which is weird.

paulidale commented:
@levitte, the `make update` changes and why they are failing are beyond my current understanding.

dlegaultbbry commented:
What's going to happen with the /dev/crypto handling? Is that gone completely in openssl 3.0? Or are these going to be converted into providers?

paulidale commented:
Engines will continue to run with 3.0 unless 3.0 is configured with the `--no-deprecated` option.
At some future point (after five+ years), engines will stop working and the functionality will need to be converted to a provider.

paulidale commented:
Figured out a much better approach -- `no-deprecate` enables `no-engine`.
This makes things much lighter in terms of changes.

The various `EVP_meth_*` functions still need to be conditioned out.

paulidale commented:
@levitte, how do I condition out the _-engines_ and _-engine_ options for the last three commands?
They are `cmp`. `dgst` and `list`.

levitte commented:
> @levitte, how do I condition out the _-engines_ and _-engine_ options for the last three commands?
> They are `cmp`. `dgst` and `list`.

I would use `OPENSSL_NO_ENGINE`

paulidale commented:
What's the embedded PERL syntax for this?

paulidale commented:
Figured it out :)

levitte commented:
> What's the embedded PERL syntax for this?

You mean for the test recipes?

> Figured it out :)

Ok cool

paulidale commented:
The Perl that needs to be embedded into the .pod.in files: `{- "old but deprecated text" -}`

levitte commented:
Why not `$disabled{engine}`?  Considering you disable `engine` in Configure when in deprecation mode, you could consistently use `OPENSSL_NO_ENGINE` in C and `$disabled{engine}` or variants thereof in Perl.

paulidale commented:
The apps/list.c changes preserve backwards compatibility: `-engines` prints a message when `OPENSSL_NO_ENGINE` is defined.  Deprecation removes this.

I'll update the documentation to use `$disabled{engine}` (except for openssl-list).

levitte commented:
Ah, fair point!  Deprecation is then more extensive than `no-engine`, that makes perfect sense.

paulidale commented:
Travis is a timeout not an error.

paulidale commented:
Ping @openssl/committers for review

paulidale commented:
Reping @openssl/committers for review

levitte commented on doc/man1/openssl-cmp.pod.in:
```diff
@@ -698,22 +698,24 @@ If not given here, the password will be prompted for if needed.
-as supported, e.g., by libp11:
-C<-key engine:pkcs11:object=my-private-key;type=private;pin-value=1234>
-
+{- if (!$disabled{"deprecated-3.0"}) {
```
There's otherwise an option to turn template output off and on:

```
{- output_off() if $disabled{"deprecated-3.0"}; "" -}

=item B<-engine> I<id>
...
...
C<-key engine:pkcs11:object=my-private-key;type=private;pin-value=1234>

{- output_on() if $disabled{"deprecated-3.0"}; "" -}
```

That way, you don't have to re-edit the text, just insert the `output_off` and `output_on` lines where needed.

paulidale replied:
Okay, changed.

levitte commented on doc/man1/openssl-dgst.pod.in:
```diff
@@ -178,14 +178,16 @@ Compute HMAC using a specific key for certain OpenSSL-FIPS operations.
-When used with the B<-engine> option, it specifies to also use
-engine I<id> for digest operations.
-
+{- if (!$disabled{"deprecated-3.0"}) {
```
Same here...

levitte commented on doc/man1/openssl-list.pod.in:
```diff
@@ -92,11 +92,15 @@ a block of multiple lines, all but the first are indented.
-
-Display a list of loaded engines.
-
+{- if (!$disabled{"deprecated-3.0"}) {
```
.... and here

levitte commented:
Er...  so this ended up just being an application change?  Ah, and a configuration change...  but I'm thinking that if we're *deprecating*, we *should* do it properly, i.e. deprecate the ENGINE API using `DEPRECATEDIN_3_0` macros.

paulidale commented:
Without the configuration change, this quickly becomes a nightmare.  With the change, the existing no-engine support takes care of removing all the engine code for us.

Any suggestion as to how deprecation macros can be added to `engine.h` while still using the no-engine configuration option?

I could add a _#warning engines are deprecated_ message to the `engine.h` header but this risks breaking applications that include but don't use it.

levitte commented:
> Any suggestion as to how deprecation macros can be added to `engine.h` while still using the no-engine configuration option?

Just add them.  I don't see why you couldn't...

paulidale commented:
The whole file is conditioned out by the `no-engine` option that gets enabled.

levitte commented:
> The whole file is conditioned out by the `no-engine` option that gets enabled.

So?  The point is that when neither `no-deprecated` nor `no-engine` has been configured, the deprecation attribute should still be added to those functions.

paulidale commented:
I realised my mistake overnight.  😞

paulidale commented:
I've run through and deprecated everything in the header file.
I've updated most of the documentation that mentions engines to say that that part is deprecated.
I've cleaned up the init and tracing related to engines.

Is this PR complete now?

levitte commented on crypto/asn1/ameth_lib.c:
```diff
@@ -7,6 +7,11 @@
+/* We need to use some engine deprecated APIs */
+#ifndef OPENSSL_NO_DEPRECATED_3_0
+# define OPENSSL_SUPPRESS_DEPRECATED
+#endif
```
I don't think you need to wrap this define.  When `OPENSSL_NO_DEPRECATED_3_0` is defined, then `OPENSSL_NO_ENGINE` is defined as well, is it not?

paulidale replied:
I started wrapping these, then decided not to.  This must have slipped through.

levitte commented on crypto/trace.c:
```diff
@@ -126,8 +126,6 @@ static const struct trace_category_st trace_categories[] = {
     TRACE_CATEGORY_(TLS_CIPHER),
     TRACE_CATEGORY_(CONF),
-    TRACE_CATEGORY_(ENGINE_TABLE),
-    TRACE_CATEGORY_(ENGINE_REF_COUNT),
```
Hmmm, shouldn't this rather be wrapped in a `#ifndef OPENSSL_NO_ENGINE`?

paulidale replied:
Do we really need to support engine tracing?  This is a new feature and engines are being deprecated.  No engine user will miss this.

Note also: the documentation mentions _OSSL_TRACE_CATEGORY_ENGINE_CONF_ which isn't defined anywhere.  I've added a fixup to remove this too.

paulidale replied:
I've added the conditionals.

Also removed the _OSSL_TRACE_CATEGORY_ENGINE_CONF_ documentation.

levitte commented on doc/man1/openssl-dgst.pod.in:
```diff
@@ -30,7 +30,9 @@ B<openssl> B<dgst>|I<digest>
 {- $OpenSSL::safe::opt_engine_synopsis -}
+{- output_off() if $disabled{"deprecated-3.0"}; "" -}
 [B<-engine_impl> I<id>]
+{- output_on() if $disabled{"deprecated-3.0"}; "" -}
```
Unfortunately, the perl bits do not eat up the line ending that follows it, so you end up getting the `-engine_impl` in a paragraph of its own (empty line before and after).  The easiest trick to resolve that is this:

```
{- output_off() if $disabled{"deprecated-3.0"}; ""
-}[B<-engine_impl> I<id>]{-
   output_on() if $disabled{"deprecated-3.0"}; "" -}
```

It's not the best-looking thing in the world, but it does the job.

For larger chunks of text that are separate paragraphs anyway, this is less of a worry.

(this is a fallacy with Text::Template, but it's otherwise simple and small and suitable for our purposes...  there are other template toolkits that do this kind of stuff better, but they are also quite large)

paulidale replied:
Done here and in openssl-list.pod.in.

levitte commented:
Travis failures seems to be a few timeouts.  Not relevant here

levitte commented:
Merged

910b71cf47266233fcb0a46d3e742e1a0834069f deprecate engines in 3.0
0a684b09d8911a5849744f2de90e81ae07bd59fb apps/list: deprecate engine support
699caa18d5f28f373fb7f5c7b946249e0be1ceb4 engine: document the engine app as deprecated
0f221d9c68b005332e21e70e7e841d021dc20498 apps: document the deprecation of the -engine option
4222682dae2efde296ba43a98673b55d0ce55d7e doc: deprecate ENGINE documentation
597f3f3ab144a4f3616073f41ee3722e1b75f2a3 Fix indentation for engine.h
cf8e8cba93776cda9a137232df58a17d7c9a8fc1  deprecate engines
92f8603537d5e883588163e04396779edb21a0b7 deprecate engines in SSL
304d070eba9dae427a5846371a4575e397f7bbbd deprecate engine tests
91512a771ac0f98b741824af9b2472d3eb754e70 deprecate engine from public header files
ad8fc6f626a6f9064383d77b102c8efcf442a9d7 apps: deprecate engines
e4468e6d8dc5a8a7e49f2ba77ed55ce2b282d4a7 deprecate engines in libcrypto
9bd8d96c390902ae3c0926c1d6c68521e48a7f7f deprecate engines in provider code
2d71c9468a338adaa5c78b9d4a1382ff8f106143 doc: document that the engine initialisation options are deprecated.
571d2c4dc70b9dffb32cf1a4969820dec6426c46 ENGINESDIR: document that this configuration is deprecated.
8b4c89f8d2d0a277ad9f66e98d8f308ebb4ef8c4 RAND: document that the ENGINE RAND override is deprecated.
1bdab93a62df26112564acc94b2a753436b9ac7b Document that the ENGINE_[sg]_ex_data() calls are reprecated.
2099f1bb6ba61e9f50193d3eab22e31bba627fb9 Document that exdata for ENGINES is deprecated.
03445677b900fb9775c6652daa7fe7943d468aa7 Document that ENGINE_add_conf_module() was deprecated.
184fb690fabe272a197f363144064141d369a192 trace: condition out engine related tracing
bb9542621170185a9a4af6d0468cf839f4512343 doc: remove unused engine tracing option
81ed433cf835bf7b47aa926735196b6948f65e95 libcrypto.num: engine deprecation updates

paulidale commented:
Thanks @levitte.
